### PR TITLE
DynamicPairlist - pair_whitelist

### LINKED ISSUE #2413: Errorneous handling of active_pair_whitelist
`Freqtradebot._active_pair_whitelist` is changed in the `process()` method while throttling: first it obtains the content of the whitelist config setting or a dynamic pairlist, then the list can be adjusted by the Edge, then it's extended with the list of pairs from the open trades.

It's okay for the trading logic, which is run in the same thread. However, it's accessed in the rpc modules as well, that typically run in different threads (for api server and telegram at least), so fetching of its value is performed independently of the transformations described above. As a result, two subsequent calls to `_rpc_whitelist()` may report completely different results: either the list from the config, or one after Edge, or one extended with trades.

All transformations of the active_pair_list in freqtradebot should be done using either local vars or another intermediate attribute, and active_pair_list should only get the resulting list after all transformations.

Btw, it's actually not active pair whitelist, but active pair list (maybe renamed during fix of this issue, to reflect its content more precisely).

### LINKED ISSUE #2414: Unclear behavior of dynamic pair list
## Step 1: Have you search for this issue before posting it?
Yes.

## Step 2: Describe your environment

  * Operating system: `macOS 10.14.6`
  * Python Version: `3.6.5`
  * CCXT version: `ccxt==1.18.1180`
  * Branch: `2019.9`
  * Last Commit ID: `6bbc0eefed8e45edb5e2ca48f9e8760551b7a395`
  
## Step 3: Describe the problem:

The underlying logic of the dynamic pair list is unclear, and the doc didn't help with this. (but except this, the doc is fantastic, thanks!)

What I would like to achieve is to use market data to create a dynamic pair list, without any hardcoded whitelist. (only the blacklist)

Pairlist config:
```json
    "pairlist": {
        "method": "VolumePairList",
        "config": {
            "number_assets": 100,
            "sort_key": "quoteVolume",
            "precision_filter": true
        }
    },
```

Since the whitelist parameter is required, I've configured an extensive list of pairs (under the `exchange` node), and I expect the `VolumePairList method` + the `precision_filter` to disable the irrelevant ones, but this isn't straightforward.

1. Is it possible to have no whitelist and to fill it dynamically from market data?

2. I've noticed a few **trades on pairs that weren't whitelisted**.
Is it because the pairs were somehow "cherry-picked" by the `VolumePairList` method?
Then probably the `pair_whitelist` parameter shouldn't be mandatory.
If this isn't the normal behavior, there is a bug that trades outside of the white list.

It seems too big to be a bug, I'm probably missing something, help is appreciated.

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary
`pair_whitelist` is not mandatory (it's not used at all) if VolumePairList is used.

Improve dynamic pairlist documentation to mention that pair_whitelist is not used and not needed by VolumePairList.

Closes #2414
closes #2413
## Quick changelog

- Lift requirement of having pair_whitelist if not using `StaticPairList`
- Update docs

### CODE REVIEW

**Comment by hroff-1902** [comment]
Hmm, I thought that the empty list `[]` stated in #2414 was due to this assignment:

https://github.com/freqtrade/freqtrade/blob/74b2f11d4f90f963c2aaf9acb65ce8cbfcf62d32/freqtrade/freqtradebot.py#L75

I've looked at this couple of days before and my opinion was that this assignment should better be equal to that in the `process()`:
```
self.active_pair_whitelist = self.pairlists.whitelist
```
-- we've already got exchange, dataprovider, strategy initialized and pairlist resolved at this point, so nothing should prevent us from fetching the whitelist at this point (maybe explicit pairlist refresh is required before...)

**Comment by xmatthias** [comment]
This got slightly out of hand on the tests ... 

because the whitelist is now refreshed on freqtrade initialization, we'll need to mock this away since otherwise the list is shuffled (by converting it to set and back), and test-results become random.

Also, i had to change which pairs are inactive. All whitelisted pairs in default_conf should be active - tests for inactive pairs did already use other pairs. However, this now created a problem, since these pairs were thrown out of the list.

File: tests/conftest.py [hunk 1]
```diff
@@ -120,6 +129,8 @@ def patch_freqtradebot(mocker, config) -> None:
     patch_exchange(mocker)
     mocker.patch('freqtrade.freqtradebot.RPCManager._init', MagicMock())
     mocker.patch('freqtrade.freqtradebot.RPCManager.send_msg', MagicMock())
+    mocker.patch('freqtrade.freqtradebot.FreqtradeBot._refresh_whitelist',
+                 MagicMock(return_value=config['exchange']['pair_whitelist']))
```
[the remark this answers is not in the document]
Comment (xmatthias) [reply tests/conftest.py]: Nice catch! :+1: 
i added this first and did not remember to replace it afterwards

**Comment by xmatthias** [comment]
The "set" conversation is here:

https://github.com/freqtrade/freqtrade/blob/7329fce87cc273c1adfe6285720b63522ce973fd/freqtrade/pairlist/IPairList.py#L63-L89

It's not an explicit shuffle, but set is an unordered structure - so it may shuffle elements.

> Being an unordered collection, sets do not record element position or order of insertion. Accordingly, sets do not support indexing, slicing, or other sequence-like behavior.

[source](https://docs.python.org/3.8/library/stdtypes.html#set-types-set-frozenset)

**Comment by hroff-1902** [comment]
> The "set" conversation is here:

Ah, I've looked at this. Should probably be implemented with OrderedList some days...

**Comment by xmatthias** [comment]
well what makes alphabetic sorting correct?

for StaticPairlist, sorting can very well be random (or should keep the orignal sorting) ... neither of which can be archived via OrderedList.

**Comment by hroff-1902** [comment]
> well what makes alphabetic sorting correct?
> 
> for StaticPairlist, sorting can very well be random (or should keep the orignal sorting) ... neither of which can be archived via OrderedList.

I assumed it should save the order of pairs from pair_whitelist (the StaticPairList pairlist, as in tests).

**Pull request merged**